src/client/httpClient.ts

```
@@ -114,12 +128,16 @@ export class HTTPClient {
     } else {
       throw new Error('baseUrl required for initialization');
     }
-
-    if (!options.applicationToken) {
-      //TODOV3 Replace this with auth call to get logged in user by providing username and password
-      throw new Error('applicationToken required for initialization');
-    } else{
+    this.username = options.username || '';
+    this.password = options.password || '';
+    this.authUrl = options.authUrl || this.baseUrl + AUTH_API_PATH;
+    if (options.applicationToken) {
       this.applicationToken = options.applicationToken;
+    }else{      
+      if(this.username === '' || this.password === ''){
```

Should also throw this error if `!this.authUrl`. If authUrl is not set, you'll end up getting a `Invalid value 'undefined' for header "X-Cassandra-Token"` error whenever you run your first operation.

---

I have a check before this condition as below to make sure the `authUrl` is always set. I believe that should take care?

`this.authUrl = options.authUrl || this.baseUrl + AUTH_API_PATH;`

---

Hi @vkarpov15 could you please approve if this looks good?

---

I took a look and my issue is due to the `authUrl` being incorrect. It looks like the `authUrl` should be `127.0.0.1:8081/v1/auth`. But the default `this.baseUrl + AUTH_API_PATH` becomes `127.0.0.1:8180/v1/auth`, different port.

Is the default value of `127.0.0.1:8180/v1/auth` expected to successfully return an auth token?

Also, would be nice to improve the error message if there's no access token. Right now, if auth fails, the error message is `Invalid value "undefined" for header "X-Cassandra-Token"`, which would be nice to improve.

---

I set `this.baseUrl + AUTH_API_PATH` as the default because, while connecting to the AstraDB we don't have to change the `baseUrl`, just appending `AUTH_API_PATH` should take care of it.
But, on local or running against open source json api with separate auth endpoint, we will have to provide the `authUrl` separately as a connection option.

---

Also, improved the error handling when no token is present. Please see commit [44f4024](https://github.com/stargate/stargate-mongoose/pull/25/commits/44f40247ed30d9cc6ff94455c94bf9d54c902143).

Please approve if this looks good.